rfcs/20190411-tensorboard-improved-plugin-ext.md

```diff
@@ -0,0 +1,151 @@
+
+A frontend binary will be a self evaluating JavaScript. With exceptions of Polymer1 and Polymer2, modern frameworks and libraries either use JavaScript or a JavaScript based template system to create a DOM or a WebComponent and, thus, the requirement should be easy to fulfill. To support legacy Polymer versions, we can consider taking a HTML as the format for the binary but this is not robust to different binary loading strategies discussed above, especially given deprecated [HTML import](https://developer.mozilla.org/en-US/docs/Web/Web_Components/HTML_Imports).
+
+### Core module and UI/UX Consistency
+One of the most important key concepts when using TensorBoard is the data selection. TensorBoard uses runs and tags to filter potentially a large set of event logs to visualize. Most plugins implemented this today and we expect third party plugins to do the same.
+
+For ease and to promote maximal interaction and UI consistency, TensorBoard will provide a library that bundles Polymer components and encourage plugin authors to use the building blocks to build out a plugin. One of the components will be tf-runs-selector which will not only provide necessary controls but also tie with the TensorBoard core and make the selection consistent across all plugins. Note that the component can be used via Polymer or by using `document.createElement('tf-runs-selector')` if using other view libraries.
+
+### Plugins Misc
+#### Accessibility
+N/A: Using an iframe does not change behavior screen reader
+#### Instrumentation Considerations
+A plugin may decide to put instrumentations and there is no way for TensorBoard to enforce/disallow that.
+#### Docs and Tutorial
+How can a plugin show a helpful manual without asking user to visit the repository to review the README? Can TensorBoard assist by showing HTMLfied README in the browser?
+
+## Alternatives considered
+### Iframe and Inter-frame communication
+Using XMLHttpRequest, TensorBoard can fetch the content of an iframe as text and use `iframe.srcdoc` to render the content. This iframe, however, will have origin of `null` and its requests will not be intercepted by any ServiceWorker.
+
+Instead, TensorBoard can provide a library one of which will be a `request`. The method will use `window.postMessage` to route the network request to the main frame and hide away the fact that it is using inter-frame communication at all. This solution, though easy for XMLHttpRequest based request, is difficult to apply for, for example, image, audio, video, CSS, web-fonts, EsModule, and WebSocket requests. To illustrate the problem more concretely, when `<img src="foo.png">` is created, a browser initiates the requests and it is not possible to route that request via the library. This will make a plugin like Images plugin difficult to write.
+
+## Questions and Discussion Topics
```

Comment:
Big missing pieces I see in this draft

1. What is the signature of `get_plugin()`, i.e., how tensorboard discovers the 3p plugin? See my comment about the "standard interface" above.
2. Apart from the standard interface, are there going to be any new protocols for any of the following?
  - data storage in the backend (e.g., tensorboard core can provide a unified database write/read API)
  - communication protocol between the plugin's frontend component and the backends (e.g., tensorboard can provide an NPM package that helps the frontend component manage the requests to the backend).

Reply:
+1

Building on the discussion of the communication protocol, Does this illustration capture how you expect the pattern of communication to work?  Or do you rather expect the TensorBoard FE to pass an address to the Plugin FE so that the Plugin FE can talk to the backend components directly?


![image](https://user-images.githubusercontent.com/547150/56048719-14bb6580-5d16-11e9-8359-b5be7b990091.png)

Reply:
@bileschi thanks for the feedback. At this moment, we would like to confine the scope of design to redo the distribution strategy. The architecture you are describing has more profound impact than depicted in the picture and should be redesigned holistically (including the summary system and storage system. Note that the storage should also be pluggable: data format, storage mechanism, and etc...).

Before closing it, I would also like to note that close coupling between frontend and backend (and distributed as one unit) alleviates the [concern of fragmentation](https://github.com/tensorflow/community/pull/90#issuecomment-482559006) raised below. Though hard to avoid duplication of features and varying level of quality, current plugin system allow us to do without API versioning and conflict resolutions.

Reply:
Thanks for the feedback!  Can you provide a substitute diagram which describes the proposed communication strategy?

Reply:
I also strongly believe that we should reconsider the architecture of providing plugins as coupled (Backend, Frontend) pairs.

At the same time, I think that problem is somewhat separable from the current proposal.  That is, this RFC goes one step in the right direction; we'd like to discuss additional related steps soon, but they don't need to block this one.

In particular I think we can easily get consensus on these bits:

* Plugins providing BE Python code are dynamically loaded at runtime.
* BE plugins are distributed via PIP.

While these aspects of the current architecture (which this RFC retains) should remain open topics:

* Plugins bundle FE code.
* The TensorBoard backend is responsible for assembling the FE app from the available plugins and serving it.
* Plugin network APIs are not formalized or versioned.

The reason this is coming up now is that most of this doc is specifically about how the plugin FEs should be assembled into the FE app that TB serves.  Under a decoupled architecture in which the FE is built and served as an independent entity (presumably, from components distributed via npm, bundled with webpack or similar, etc.), the discussion around whether or not to use iFrames etc. would be very different.

The upshot is I think it's OK to go ahead with this proposal, because the dynamic BE plugin loading (and PIP distribution of those) is great in any case.  It may turn out that the FE-assembly questions here will be obviated by a later design of broader scope; let's give that the attention it deserves in a separate RFC.

Reply:
@bileschi As I understand it, the current proposal does not change the existing communication strategy, which is that a plugin FE makes requests only to routes that its matched BE exposes (or to standard routes that the TB core provides).  See https://github.com/tensorflow/tensorboard/blob/master/tensorboard/http_api.md.  There is no client-side TB code mediating those requests, and no client-side state shared between FE plugins.  (Those are certainly good things to discuss; they're just outside the present scope.)